File: src/main/java/de/churl/cookbook/service/PersistenceService.java
package de.churl.cookbook.service;

import de.churl.cookbook.exception.IngredientNotFoundException;
import de.churl.cookbook.exception.RecipeNotFoundException;
import de.churl.cookbook.model.Ingredient;
import de.churl.cookbook.model.IngredientType;
import de.churl.cookbook.model.IngredientUsage;
import de.churl.cookbook.model.IngredientUsageKey;
import de.churl.cookbook.model.Recipe;
import de.churl.cookbook.model.transfer.IngredientDTO;
import de.churl.cookbook.model.transfer.RecipeDTO;
import de.churl.cookbook.persistance.IngredientRepository;
import de.churl.cookbook.persistance.RecipeRepository;
import de.churl.cookbook.utility.IdUtil;
import de.churl.cookbook.utility.TransferUtil;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.Optional;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class PersistenceService {

    private final RecipeRepository recipeRepository;
    private final IngredientRepository ingredientRepository;

    // Recipe ##########################################################################################################

    public Iterable<RecipeDTO> findAllRecipes() {
        return TransferUtil.recipeToDTO(this.recipeRepository.findAll());
    }

    public RecipeDTO findRecipeById(String id) {
        return this.findRecipeById(IdUtil.toUUID(id));
    }

    public RecipeDTO findRecipeById(UUID id) {
        return TransferUtil.recipeToDTO(this.recipeRepository.findById(id).orElseThrow(() -> new RecipeNotFoundException(id)));
    }

    /**
     * Saves a new recipeDTO in the database.
     * The new entity is constructed with a random UUID.
     *
     * @param title The recipeDTO title
     * @param descr The recipeDTO description
     * @param body The recipeDTO body
     * @param ingrs The needed ingrDTOs (List of IDs)
     * @return The id of the saved entity
     */
    public UUID saveNewRecipe(String title, String descr, String body, Collection<UUID> ingrs) {
        final Recipe recipe = new Recipe();
        recipe.setRecipeID(UUID.randomUUID());
        recipe.setRecipeTitle(title);
        recipe.setRecipeDescr(descr);
        recipe.setRecipeBody(body);
        recipe.setIngrUsages(new HashSet<>());

        for (UUID ingr : ingrs) {
            final Optional<Ingredient> ingropt = this.ingredientRepository.findById(ingr);
            final Ingredient ingredient = ingropt.orElseThrow(() -> new IngredientNotFoundException(ingr));
            final IngredientUsage usage = new IngredientUsage();
            final IngredientUsageKey key = new IngredientUsageKey();

            key.setRecipeID(recipe.getRecipeID());
            key.setIngrID(ingr);

            usage.setIngredientUsageKey(key);
            usage.setRecipe(recipe);
            usage.setIngr(ingredient);
            usage.setAmount(1); // TODO: Ingredient Amount
            recipe.getIngrUsages().add(usage);
        }

        return this.recipeRepository.save(recipe).getRecipeID();
    }

    // Ingredient ######################################################################################################

    public Collection<IngredientDTO> findAllIngredients() {
        return TransferUtil.ingrToDTO(this.ingredientRepository.findAll());
    }

    public IngredientDTO findIngredientById(String id) {
        return this.findIngredientById(IdUtil.toUUID(id));
    }

    public IngredientDTO findIngredientById(UUID id) {
        return TransferUtil.ingrToDTO(this.ingredientRepository.findById(id).orElseThrow(() -> new IngredientNotFoundException(id)));
    }

    /**
     * Saves a new ingredient in the database.
     * The new entity is constructed with a random UUID.
     * The IngredientUsages-Set is initialized empty.
     *
     * @param title The ingredient title
     * @param type The ingredient type
     * @return The id of the saved entity
     */
    public UUID saveNewIngredient(String title, String type) {
        final Ingredient ingr = new Ingredient();
        ingr.setIngrID(UUID.randomUUID());
        ingr.setIngrTitle(title);
        ingr.setIngrType(IngredientType.valueOf(type));
        ingr.setIngrUsages(Collections.emptySet());

        return this.ingredientRepository.save(ingr).getIngrID();
    }
}

File: src/main/kotlin/de/churl/cookbook/service/PersistenceService.kt
package de.churl.cookbook.service

import de.churl.cookbook.error.IngredientNotFoundError
import de.churl.cookbook.error.RecipeNotFoundError
import de.churl.cookbook.model.*
import de.churl.cookbook.model.transfer.IngredientDTO
import de.churl.cookbook.model.transfer.RecipeDTO
import de.churl.cookbook.persistence.IngredientRepository
import de.churl.cookbook.persistence.RecipeRepository
import de.churl.cookbook.util.ingrToDTO
import de.churl.cookbook.util.recipeToDTO
import de.churl.cookbook.util.toUUID
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import java.util.*

@Service
class PersistenceService(
    val ingrRepository: IngredientRepository,
    val recipeRepository: RecipeRepository
) {
    // FIND RECIPES ################################################################################

    fun findAllRecipes(): Map<UUID, RecipeDTO> {
        return recipeRepository.findAll().associateBy({ it.recipeID }, { recipeToDTO(it) })
    }

    fun findRecipeById(id: String): RecipeDTO {
        return findRecipeById(toUUID(id))
    }

    fun findRecipeById(id: UUID): RecipeDTO {
        return recipeToDTO(recipeRepository.findByIdOrNull(id) ?: throw RecipeNotFoundError(id))
    }

    // SAVE RECIPES ################################################################################

    fun saveNewRecipe(recipe: RecipeDTO): UUID {
        return saveRecipe(
            recipe.title,
            recipe.descr,
            recipe.body,
            toUUID(recipe.ingrs),
            null
        )
    }

    fun saveRecipe(recipe: RecipeDTO, id: UUID): UUID {
        return saveRecipe(
            recipe.title,
            recipe.descr,
            recipe.body,
            toUUID(recipe.ingrs),
            id
        )
    }

    fun saveRecipe(title: String, descr: String, body: String, ingrs: Collection<UUID>, id: UUID?): UUID {
        val recipe = Recipe(
            title,
            descr,
            body
        )

        for (id: UUID in ingrs) {
            val ingr = ingrRepository.findByIdOrNull(id) ?: throw IngredientNotFoundError(id)
            val ingrUsageKey = IngrUsageKey(recipe.recipeID, ingr.ingrID)
            val ingrUsage = IngrUsage(ingrUsageKey, recipe, ingr, "1")

            recipe.ingrUsages.add(ingrUsage)
        }

        if (id != null) recipe.recipeID = id

        return recipeRepository.save(recipe).recipeID
    }

    // FIND INGRS ##################################################################################

    fun findAllIngredients(): Map<UUID, IngredientDTO> {
        return ingrRepository.findAll().associateBy({ it.ingrID }, { ingrToDTO(it) })
    }

    fun findAllIngredientsById(ids: Collection<String>): Collection<IngredientDTO> {
        return ingrToDTO(ingrRepository.findAllById(toUUID(ids)))
    }

    fun findIngrById(id: String): IngredientDTO {
        return findIngrById(toUUID(id))
    }

    fun findIngrById(id: UUID): IngredientDTO {
        return ingrToDTO(ingrRepository.findByIdOrNull(id) ?: throw IngredientNotFoundError(id))
    }

    // SAVE INGRS ##################################################################################

    fun saveNewIngredient(ingr: IngredientDTO): UUID {
        return saveNewIngredient(ingr.title, ingr.type)
    }

    fun saveNewIngredient(title: String, type: String): UUID {
        val ingr = Ingredient(
            title,
            IngredientType.valueOf(type)
        )

        return ingrRepository.save(ingr).ingrID
    }
}
